launch_testing/test/launch_testing/test_proc_info_assertions.py

```diff
@@ -0,0 +1,128 @@
+                        )
+                    )
+                ]
+            ),
+            launch.actions.OpaqueFunction(function=lambda context: ready_fn())
+        ]), {'good_process': good_process}
+
+    # This is kind of a weird test-within-a-test, but it's the easiest way to get
+    # all of the proc_info handlers hooked up correctly without digging deep into
+    # the guts of the runner
+    def test_01_check_running_process(self, proc_info, good_process):
+        with self.assertRaisesRegex(AssertionError, 'Timed out waiting for process'):
+            proc_info.assertWaitForShutdown(good_process, timeout=3)  # Should raise
+
+    def test_02_check_when_process_exits(self, proc_info, good_process):
+        proc_info.assertWaitForShutdown(good_process, timeout=15)
+
+    runner = LaunchTestRunner(
+        source_test_loader(
+            generate_test_description,
+            pre_shutdown_tests=[
+                test_01_check_running_process,
+                test_02_check_when_process_exits,
+            ]
+        )
+    )
+    results = runner.run()
+
+    for result in results.values():
+        assert result.wasSuccessful()
+
+
+def test_wait_for_startup(source_test_loader):
+
+    def generate_test_description(ready_fn):
+        TEST_PROC_PATH = os.path.join(
+            ament_index_python.get_package_prefix('launch_testing'),
+            'lib/launch_testing',
+            'good_proc'
+        )
+
+        good_process = launch.actions.ExecuteProcess(
+                cmd=[sys.executable, TEST_PROC_PATH],
+        )
+
+        # Start 'good_proc' after a ten second timer elapses
+        return launch.LaunchDescription([
+            launch.actions.TimerAction(
+                period=10.0,
+                actions=[good_process]
+            ),
+            launch.actions.OpaqueFunction(function=lambda context: ready_fn())
+        ]), {'good_process': good_process}
+
+    # This is kind of a weird test-within-a-test, but it's the easiest way to get
+    # all of the proc_info handlers hooked up correctly without digging deep into
+    # the guts of the runner
+    def test_01_check_for_non_running_process(self, proc_info, good_process):
+        with self.assertRaisesRegex(AssertionError, 'Timed out waiting for process'):
+            proc_info.assertWaitForStartup(good_process, timeout=3)  # Should raise
+
+    def test_02_check_when_process_runs(self, proc_info, good_process):
+        # The process we're waiting for should start about 7 seconds into this wait
+        proc_info.assertWaitForStartup(good_process, timeout=15)
+
+    runner = LaunchTestRunner(
+        source_test_loader(
+            generate_test_description,
+            pre_shutdown_tests=[
+                test_01_check_for_non_running_process,
+                test_02_check_when_process_runs
+            ]
+        )
+    )
+    results = runner.run()
+
+    for result in results.values():
+        assert result.wasSuccessful()
```

Comment:
Hmm, on a second look this seems like a launch test fit into a `pytest`. It's fine for now but we should probably standardize into something less terse and/or use launch test where appropriate.

Reply:
@hidmic You're not wrong - line 58/105 explains what's going on.  I want to use the runner here because plumbing the proc_info handler to the launch is a bit complicated and would be prone to errors if I did it in the test.

If you think it would be more clear to make these into launch_tests, I can try to do it that way.  Making fake Launch events is a bit difficult - I felt it was less error prone to use 'launch' to make real events

Reply:
No, fake `launch` events doesn't sound great either. This likely falls in the bucket of things we need to iterate on as a whole to provide a better API to our users.